File: docs/public/specs/esm/density1d.js
import * as vg from "@uwdata/vgplot";

await vg.coordinator().exec([
  vg.loadParquet("flights", "data/flights-200k.parquet")
]);

const $brush = vg.Selection.crossfilter();
const $bandwidth = vg.Param.value(10);

export default vg.vconcat(
  vg.slider({label: "Bandwidth (σ)", as: $bandwidth, min: 0.1, max: 100, step: 0.1}),
  vg.plot(
    vg.densityY(
      vg.from("flights", {filterBy: $brush}),
      {x: "delay", fill: "#888", fillOpacity: 0.5, bandwidth: $bandwidth}
    ),
    vg.intervalX({as: $brush}),
    vg.yAxis(null),
    vg.xDomain(vg.Fixed),
    vg.width(600),
    vg.marginLeft(10),
    vg.height(200)
  ),
  vg.plot(
    vg.densityY(
      vg.from("flights", {filterBy: $brush}),
      {x: "distance", fill: "#888", fillOpacity: 0.5, bandwidth: $bandwidth}
    ),
    vg.intervalX({as: $brush}),
    vg.yAxis(null),
    vg.xDomain(vg.Fixed),
    vg.width(600),
    vg.marginLeft(10),
    vg.height(200)
  )
);

File: specs/python/density1d.py
import vgplot as vg

meta = vg.meta(
    title="Density 1D",
    description="Density plots (`densityY` mark) for over 200,000 flights, created using kernel density estimation. Binning is performned in-database, subsequent smoothing in-browser. The distance density uses a log-scaled domain. To change the amount of smoothing, use the slider to set the kernel bandwidth.\n",
)
flights = vg.parquet("data/flights-200k.parquet")

brush = vg.selection.crossfilter()
bandwidth = vg.param(20)

view = vg.vconcat(
    vg.slider(label="Bandwidth (σ)", bind=bandwidth, min=0.1, max=100, step=0.1),
    vg.plot(
        vg.density_y(
            data="flights",
            filter_by=brush,
            x="delay",
            fill="#888",
            fill_opacity=0.5,
            bandwidth=bandwidth,
        ),
        vg.interval_x(bind=brush),
        vg.y_axis(None),
        vg.x_domain("Fixed"),
        vg.width(600),
        vg.margin_left(10),
        vg.height(200),
    ),
    vg.plot(
        vg.density_y(
            data="flights",
            filter_by=brush,
            x="distance",
            fill="#888",
            fill_opacity=0.5,
            bandwidth=bandwidth,
        ),
        vg.interval_x(bind=brush),
        vg.y_axis(None),
        vg.x_scale("log"),
        vg.x_domain("Fixed"),
        vg.width(600),
        vg.margin_left(10),
        vg.height(200),
    ),
)

spec = vg.spec()
